### PR TITLE
Vas 29

### PR DESCRIPTION
**Pull request by ylatuya**

### CODE REVIEW

File: VAS.Core/Events/EventsBroker.cs [hunk 1]
```diff
@@ -39,62 +40,100 @@ public class EventsBroker
...
-		public EventToken Subscribe<TEvent> (Action<TEvent> action, ThreadMethod threadOption = ThreadMethod.PublisherThread, bool keepSubscriberReferenceAlive = false, Predicate<TEvent> filter = null)
+		/// <summary>
+		/// Subscribes to an event asynchronously.
+		/// </summary>
+		/// <param name="action">Callback function called when the event is raised.</param>
+		/// <param name="threadOption">Thread option.</param>
+		/// <param name="keepSubscriberReferenceAlive">If set to <c>true</c> keep subscriber reference alive.</param>
+		/// <param name="filter">Filter.</param>
+		/// <typeparam name="TEvent">The type of the event.</typeparam>
+		public EventToken SubscribeAsync<TEvent> (Func<TEvent, Task> action, ThreadMethod threadOption = ThreadMethod.PublisherThread, bool keepSubscriberReferenceAlive = false, Predicate<TEvent> filter = null)
```
Comment (vguzmanp) [inline VAS.Core/Events/EventsBroker.cs]: What's the difference between the two subscribe methods? They use the same code

File: VAS.Core/Prism/AsyncEventSubscription.cs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+		///</summary>
+		///<param name="actionReference">A reference to a delegate of type <see cref="System.Action"/>.</param>
+		///<exception cref="ArgumentNullException">When <paramref name="actionReference"/> or <see paramref="filterReference"/> are <see langword="null" />.</exception>
+		///<exception cref="ArgumentException">When the target of <paramref name="actionReference"/> is not of type <see cref="System.Action"/>.</exception>
+		public AsyncEventSubscription (IDelegateReference actionReference)
+		{
+			if (actionReference == null)
+				throw new ArgumentNullException (nameof (actionReference));
+			if (!(actionReference.Target is Action))
+				throw new ArgumentException (string.Format (CultureInfo.CurrentCulture, Resources.InvalidDelegateRerefenceTypeException, typeof (Action).FullName), nameof (actionReference));
```
Comment (vguzmanp) [inline VAS.Core/Prism/AsyncEventSubscription.cs]: Add braces

File: VAS.Core/Prism/AsyncEventSubscription.cs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+		/// <summary>
+		/// Gets the target <see cref="System.Threading.Tasks"/> that is referenced by the <see cref="IDelegateReference"/>.
+		/// </summary>
+		/// <value>An <see cref="System.Action"/> or <see langword="null" /> if the referenced target is not alive.</value>
+		public Func<Task> Action {
+			get { return (Func<Task>)_actionReference.Target; }
+		}
+
+		public Delegate Delegate {
```
Comment (vguzmanp) [inline VAS.Core/Prism/AsyncEventSubscription.cs]: Add documentation

File: VAS.Core/Prism/AsyncEventSubscription.cs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+		}
+
+		/// <summary>
+		/// Invokes the specified <see cref="System.Action{TPayload}"/> synchronously when not overridden.
+		/// </summary>
+		/// <param name="action">The action to execute.</param>
+		/// <exception cref="ArgumentNullException">An <see cref="ArgumentNullException"/> is thrown if <paramref name="action"/> is null.</exception>
+		public virtual Task InvokeAction (Func<Task> action)
+		{
+			if (action == null) throw new ArgumentNullException (nameof (action));
```
Comment (vguzmanp) [inline VAS.Core/Prism/AsyncEventSubscription.cs]: Add braces

File: VAS.Tests/Events/TestEventsBroker.cs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+				return AsyncHelpers.Return ();
+			};
+			Action<ReturningValueEvent> act2 = (a) => {
+				count++;
+			};
+
+			// Act
+			eventsBroker.SubscribeAsync (act1);
+			eventsBroker.Subscribe (act2);
+			await eventsBroker.Publish (new ReturningValueEvent ());
```
Comment (vguzmanp) [inline VAS.Tests/Events/TestEventsBroker.cs]: How will this behave with async handlers that take long to execute? Do the handlers await between them, or they all run instantly and finish at different times?

File: VAS.Core/Events/EventsBroker.cs [hunk 6]
```diff
@@ -38,16 +38,11 @@ public class EventToken
...
-			get {
-				return _current ?? (_current = new EventAggregator ());
-			}
-			set {
-				_current = value;
-			}
-		}
+			get;
+			set;
+		} = new EventAggregator ();
```
Comment (vguzmanp) [inline VAS.Core/Events/EventsBroker.cs]: The previous version was a singleton (although not matching our coding style). This one isn't singleton. Which one do we want?

File: VAS.Core/Events/EventsBroker.cs [hunk 7]
```diff
@@ -64,6 +64,18 @@ public Task Publish<TEvent> (TEvent @event)
 		}
 
 		/// <summary>
+		/// Helper method to publish a new <see cref="ReturningValueEvent"/> and return the return value set in the event.
+		/// </summary>
+		/// <returns>The return value of the event.</returns>
+		/// <param name="event">Event.</param>
+		/// <typeparam name="TEvent">The type of the event.</typeparam>
+		public async Task<bool> CheckPublish<TEvent> (TEvent @event) where TEvent : ReturningValueEvent
```
Comment (vguzmanp) [inline VAS.Core/Events/EventsBroker.cs]: I don't really like this name `CheckPublish` sounds like it's checking what's published, but not publishing.

File: VAS.Core/Common/Enums.cs [hunk 8]
```diff
@@ -151,8 +152,6 @@ public enum EditorState
 	{
 		START = 0,
 		FINISHED = 1,
-		CANCELED = -1,
-		ERROR = -2
```
Comment (vguzmanp) [inline VAS.Core/Common/Enums.cs]: Is this related?

File: VAS.Core/Common/Job.cs [hunk 9]
```diff
@@ -19,76 +19,105 @@
...
+	/// A job running in the background.
+	/// </summary>
+	public class Job : BindableBase
 	{
 		public Job (EncodingSettings encSettings)
 		{
 			EncodingSettings = encSettings;
-			State = JobState.NotStarted;
+			State = JobState.None;
+			Progress = -1;
```
Comment (vguzmanp) [inline VAS.Core/Common/Job.cs]: Shouldn't `Progress` initialize to 0?

File: VAS.Tests/Events/TestEventsBroker.cs [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+		[SetUp]
+		public void SetUp ()
+		{
+			eventsBroker = new EventsBroker ();
+		}
+
+		[TearDown]
+		public void TearDown ()
+		{
+			SetupClass.Initialize ();
```
Comment (ylatuya) [inline VAS.Tests/Events/TestEventsBroker.cs]: This test touches things in App.Current, SetupClass.Initialize () reset everything for other tests

File: VAS.Tests/Events/TestEventsBroker.cs (hunk 5 above)
Comment (ylatuya) [inline VAS.Tests/Events/TestEventsBroker.cs]: They run sequentially, awaited between them

File: VAS.Core/Common/Enums.cs (hunk 8 above)
Comment (ylatuya) [inline VAS.Core/Common/Enums.cs]: Yes, it's part of the cleanup in the jobs manager. Errors and cancellation should not be notified using the progress, it's old behaviour that's not longer needed

File: VAS.Core/Common/Job.cs (hunk 9 above)
Comment (ylatuya) [inline VAS.Core/Common/Job.cs]: I choosed -1 because 0 can also mean the task has started and it's a 0%

File: VAS.Core/VAS.Core.Net45.csproj [hunk 11]
```diff
@@ -48,9 +48,6 @@
     <Reference Include="PropertyChanged">
       <HintPath>$(SolutionDir)\packages\PropertyChanged.Fody.1.50.3\Lib\portable-net4+sl4+wp8+win8+wpa81+MonoAndroid16+MonoTouch40\PropertyChanged.dll</HintPath>
     </Reference>
-    <Reference Include="Prism">
-      <HintPath>$(SolutionDir)\packages\Prism.Core.6.1.0\lib\net45\Prism.dll</HintPath>
-    </Reference>
```
Comment (ylatuya) [inline VAS.Core/VAS.Core.Net45.csproj]: We don't use anything else from Prism

File: VAS.Core/Prism/PubSubEvent.cs [hunk 12]
```diff
@@ -1,293 +1,386 @@
...
+		/// <summary>
+		/// Subscribes a delegate to an event that will be published on the <see cref="ThreadOption.PublisherThread"/>.
+		/// <see cref="PubSubEvent"/> will maintain a <see cref="WeakReference"/> to the target of the supplied <paramref name="action"/> delegate.
+		/// </summary>
+		/// <param name="action">The delegate that gets executed when the event is published.</param>
+		/// <returns>A <see cref="SubscriptionToken"/> that uniquely identifies the added subscription.</returns>
+		/// <remarks>
+		/// The PubSubEvent collection is thread-safe.
+		/// </remarks>
+		public SubscriptionToken Subscribe (Action action)
```
Comment (ylatuya) [inline VAS.Core/Prism/PubSubEvent.cs]: That's how Prism does it, instead of using conditional parameter it creates extra contructors

File: VAS.Core/Events/EventsBroker.cs (hunk 7 above)
Comment (vguzmanp) [inline VAS.Core/Events/EventsBroker.cs]: `PublishReturning`? `PublishAndCheck`?

File: VAS.Core/Interfaces/IJobsManager.cs [hunk 13]
```diff
@@ -20,27 +20,35 @@
...
+		/// Retry the list of jobs.
+		/// </summary>
+		/// <param name="jobs">Jobs.</param>
+		void Retry (List<Job> jobs);
+
+		/// <summary>
+		/// Cancel the list of jobs. If <paramref name="jobs"/> is <c>null</c> cancell the current job.
+		/// </summary>
+		/// <param name="cancelJobs">Cancel jobs.</param>
+		void Cancel (List<Job> jobs = null);
```
Comment (vguzmanp) [inline VAS.Core/Interfaces/IJobsManager.cs]: Better to use `IEnumerable`

File: VAS.Services/RenderingJobsController.cs [hunk 14]
```diff
@@ -25,140 +25,173 @@
...
 
-		public bool Disposed { get; private set; } = false;
+		public RenderingJobsController (JobsManagerVM viewModel)
+		{
+			ViewModel = viewModel;
+		}
+
+		protected override void Dispose (bool disposing)
+		{
+			CancelCurrentJob (false);
```
Comment (vguzmanp) [inline VAS.Services/RenderingJobsController.cs]: We don't call the `base`?

File: VAS.Services/RenderingJobsController.cs [hunk 15]
```diff
@@ -25,140 +25,173 @@
...
+			return true;
 		}
 
-		public void ClearDoneJobs ()
+		public void HandleAddJob (CreateEvent<Job> evt)
 		{
-			jobs.Model.RemoveAll (j => j.State == JobState.Finished);
+			Job job = evt.Object;
+			if (job == null)
+				return;
```
Comment (vguzmanp) [inline VAS.Services/RenderingJobsController.cs]: Add braces

Comment (vguzmanp) [inline VAS.Services/RenderingJobsController.cs]: Shouldn't this handler be the one creating the jobs?

File: VAS.Services/RenderingJobsController.cs [hunk 16]
```diff
@@ -321,153 +354,56 @@ private string CreateStillImage (MediaFile file, FrameDrawing drawing)
...
-		}
-
-		private void StartNextJob ()
-		{
-			if (pendingJobs.Count () == 0) {
-				ResetGui ();
+			CleanVideoEditor ();
+			JobVM nextJob = ViewModel.PendingJobs.FirstOrDefault ();
+			if (nextJob == null) {
+				ViewModel.CurrentJob.Model = null;
```
Comment (vguzmanp) [inline VAS.Services/RenderingJobsController.cs]: Shouldn't we work with `CurrentJob` ? Why do we need to access the Model?

File: VAS.Tests/Services/TestRenderingJobsController.cs [hunk 17]
```diff
@@ -15,47 +15,50 @@
...
 				File.Delete (file1);
 				File.Delete (file2);
 			} catch {
 			}
+			SetupClass.Initialize ();
 		}
 
 		[SetUp]
-		public void SetupMocks ()
+		public void Setu ()
```
Comment (vguzmanp) [inline VAS.Tests/Services/TestRenderingJobsController.cs]: typo `Setu`

File: VAS.UI.Gtk2/UI/Component/RenderingStateBarView.cs [hunk 18]
```diff
@@ -62,23 +64,33 @@ public RenderingStateBarView ()
...
 		}
 
 		void HandlePropertyChanged (object sender, PropertyChangedEventArgs e)
 		{
-			SyncVMValues ();
+			SyncVMValues (e.PropertyName);
 		}
 
-		void SyncVMValues ()
+		void SyncVMValues (string propertyName = null)
```
Comment (vguzmanp) [inline VAS.UI.Gtk2/UI/Component/RenderingStateBarView.cs]: Is this ever used with `null`?

File: VAS.UI.Gtk2/gtk-gui/VAS.UI.UI.TreeView.JobsTreeView.cs [hunk 19]
```diff
@@ -0,0 +1,10 @@
+
+namespace VAS.UI.UI.TreeView
+{
+	public partial class JobsTreeView
+	{
+		private void Build ()
+		{
+		}
```
Comment (vguzmanp) [inline VAS.UI.Gtk2/gtk-gui/VAS.UI.UI.TreeView.JobsTreeView.cs]: Empty JobsTreeView?

File: VAS.Tests/Events/TestEventsBroker.cs (hunk 10 above)
Comment (ylatuya) [inline VAS.Tests/Events/TestEventsBroker.cs]: That's why we use the Setup fixture, to initialize things so tests don't have to care about it. If a test changes for some reason the App for it's onw testing purpose it should clean after itself.

File: VAS.UI.Gtk2/gtk-gui/VAS.UI.UI.TreeView.JobsTreeView.cs (hunk 19 above)
Comment (ylatuya) [inline VAS.UI.Gtk2/gtk-gui/VAS.UI.UI.TreeView.JobsTreeView.cs]: This files should be deleted

File: VAS.Services/RenderingJobsController.cs (hunk 15 above)
Comment (ylatuya) [inline VAS.Services/RenderingJobsController.cs]: The Job is created by someone else, this API is AddJob () which adds a job to the queue.

File: VAS.Core/Common/Job.cs (hunk 9 above)
Comment (ylatuya) [inline VAS.Core/Common/Job.cs]: Yes, but Progress isn't 0 either, so that's why it's initialized to -1

File: VAS.Core/Prism/AsyncEventSubscription.cs [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+		/// <remarks>
+		/// If <see cref="Action"/>is no longer valid because it was
+		/// garbage collected, this method will return <see langword="null" />.
+		/// Otherwise it will return a delegate that evaluates the <see cref="Filter"/> and if it
+		/// returns <see langword="true" /> will then call <see cref="InvokeAction"/>. The returned
+		/// delegate holds a hard reference to the <see cref="Action"/> target
+		/// <see cref="Delegate">delegates</see>. As long as the returned delegate is not garbage collected,
+		/// the <see cref="Action"/> references delegates won't get collected either.
+		/// </remarks>
+		public virtual Func<object [], Task> GetExecutionStrategy ()
```
Comment (ylatuya) [inline VAS.Core/Prism/AsyncEventSubscription.cs]: Nope, because this is the non-generic AsyncEventSubscription, and there is nothing to filter against.

File: VAS.Tests/Events/TestEventsBroker.cs (hunk 10 above)
Comment (vguzmanp) [inline VAS.Tests/Events/TestEventsBroker.cs]: :+1:
